Rpi 4.4.y dsi stub squash

anholt opened the pull request:
Here's the stripped down series I've come up with for adding DSI support, based on not having ignore_lcd in the config.txt.  Since it's requiring that the firmware set up DSI, there's no power management and instead of turning off the DSI and PV during DPMS, we just set the HVS to scan out black.

There's no touch input support yet.  I think that's the next step, even before fixing the DSI link management.

mripard committed:
drm/fb_cma_helper: Remove implicit call to disable_unused_functions

The drm_fbdev_cma_init function always calls the
drm_helper_disable_unused_functions. Since it's part of the usual probe
process, all the drivers using that helper will end up having their encoder
and CRTC disable functions called at probe if their device has not been
reported as enabled.

This could be fixed by reading out from the registers the current state of
the device if it is enabled, but even that will not handle the case where
the device is actually disabled.

Moreover, the drivers using the atomic modesetting expect that their enable
and disable callback to be called when the device is already enabled or
disabled (respectively).

We can however fix this issue by moving the call to
drm_helper_disable_unused_functions out of drm_fbdev_cma_init and make the
drivers needing it (all the drivers calling drm_fbdev_cma_init and not
using the atomic modesetting) explicitly call it.

Signed-off-by: Maxime Ripard <maxime.ripard@free-electrons.com>
Link: http://patchwork.freedesktop.org/patch/msgid/1452785109-6172-14-git-send-email-maxime.ripard@free-electrons.com
Acked-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Signed-off-by: Daniel Vetter <daniel.vetter@ffwll.ch>
(cherry picked from commit 4314e19ef4ae0ba8872bd8610f6fef5e8743e236)

anholt committed:
drm/vc4: Add DSI1 driver

The DSI0 and DSI1 blocks on the 2835 are different but very similar
hardware blocks.  Some registers move around, and the featureset is
slightly different, but they're clearly related.  This doesn't enable
DSI0, but some of the infrastructure is present.

Also, this driver doesn't initialize the DSI successfully from
poweron, so we currently require that the Raspberry Pi firmware enable
it at boot time.  From there, we just keep the same settings forever,
and when poweroff is requested we just scan out black instead.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
drm/vc4: Add support for feeding DSI encoders from the pixel valve.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
drm/vc4: Fix flipped HVS channels for DSI0/1.

Empirically, DSI1 appears to be using HVS channel 0 on my Pi2.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
clk: bcm2835: Don't rate change PLLs on behalf of dividers.

Our core PLLs are intended to be configured once and left alone.  With
the flag set, asking to set the PLLD_DSI1 clock rate would change PLLD
just to get closer to the requested DSI clock, thus changing PLLD_PER,
the UART and ethernet PHY clock rates downstream of it, and breaking
ethernet.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
clk: bcm2835: Do appropriate name lookups for DSI1's parents as well.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
clk: bcm2835: Add an enum for the DSI1 pixel clock.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
drm/panel: Add support for the Raspberry Pi 7" Touchscreen.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
drm/vc4: Start switching to using debugfs_reg32 helpers for debugfs.

Every file was defining its own little struct and dumping for the
regs, when there's a helper in debugfs for doing just this.  However,
instead of printing:

PV_HORZA (0x000c): 0x00000000

we now print:

PV_HORZA = 0x00000000

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
clk: bcm2835: Mark the CM SDRAM clock's parent as critical

While the SDRAM is being driven by its dedicated PLL most of the time,
there is a little loop running in the firmware that periodically turns
on the CM SDRAM clock (using its pre-initialized parent) and switches
SDRAM to using the CM clock to do PVT recalibration.

This avoids system hangs if we choose SDRAM's parent for some other
clock, then disable that clock.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
BCM270X: Add the DSI panel to the defconfig.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
BCM270X: Add the DSI and DSI panel nodes to the VC4 overlay.

Signed-off-by: Eric Anholt <eric@anholt.net>

clivem commented on arch/arm/boot/dts/bcm2708_common.dtsi:
```diff
@@ -110,9 +111,10 @@
 			/* CPRMAN derives everything from the platform's
-			 * oscillator.
+			 * oscillator except for a few clocks that may
++                        * derive from something else derived from CPRMAN..
```
Cut and paste??????

anholt commented:
It looks like without disable_touchscreen set in config.txt, the rpi-ft5406 driver keeps working just fine.

anholt commented:
(force push was to drop the stray "+" in the comment)

chbae commented:
@anholt When will you merge this change?

anholt commented:
@chbae I'm the one requesting the merge, it's up to the foundation folks to pull it.

clivem commented:
@anholt Eric, is there no way of making this fly with CONFIG_DRM=m and CONFIG_DRM_PANEL_RASPBERRYPI_TOUCHSCREEN=m. I haven't checked, but I'm thinking that having DRM compiled-in rather than module, is going to increase the size of the default kernel substantially, when very few people are actually using the functionality...... Just thinking out loud......

tiwai committed:
drm/mipi-dsi: Unregister bus at exit

This is a preliminary patch for building drm-mipi-dsi as a module.
Add the module exit callback to unregister the bus properly.

Suggested-by: Thierry Reding <treding@nvidia.com>
Signed-off-by: Takashi Iwai <tiwai@suse.de>
Signed-off-by: Eric Anholt <eric@anholt.net>

tiwai committed:
drm/mipi-dsi: Allow to build drm-mipi-dsi as a module

The drm-mipi-dsi driver has been only built-in although this isn't
strictly required to be so.  Since it's referred by lots of DRM
drivers nowadays, most of distro kernels include the driver as
built-in as a result, even though many systems don't need it at all.

This patch fixes Kconfig to allow drm-mipi-dsi driver built as a
module, so that we can save footprint on systems without such DRM
drivers.  The probe order is managed by the module dependency, and
postcore_initcall() works just fine as a module init call.

Signed-off-by: Takashi Iwai <tiwai@suse.de>
Signed-off-by: Eric Anholt <eric@anholt.net>

popcornmix commented:
Looks like the `=y` option on CONFIG_DRM costs us 224K is available memory and 165K in size of kernel7.img.
@anholt - is that something that is essential?
It seems expensive for the users who don't use it.

```
pi@domnfs:~ $ ls -l /boot/kernel7.img 
-rwxr-xr-x 1 root root 4225384 Aug  8 18:01 /boot/kernel7.img
pi@domnfs:~ $ free -k
             total       used       free     shared    buffers     cached
Mem:        817828      91668     726160       5932         24      50740
-/+ buffers/cache:      40904     776924
Swap:       524284          0     524284
```

```
pi@domnfs:~ $ ls -l /boot/kernel7.img 
-rwxr-xr-x 1 root root 4390520 Aug  8 18:05 /boot/kernel7.img
pi@domnfs:~ $ free -k
             total       used       free     shared    buffers     cached
Mem:        817456      91544     725912       5932         24      50764
-/+ buffers/cache:      40756     776700
Swap:            0          0          0
```

anholt commented:
I don't think counting on-disk size is appropriate, as you're just moving it from an initrd module to built-in.  Memory size in non-vc4 mode is unfortunate, though if we can flip to vc4-by-defaut we should be able to get way more back by reducing gpu_mem.

For getting back to building as a module, hopefully with this series http://www.spinics.net/lists/dri-devel/msg114234.html it will be an option.

anholt commented:
Note, if you're not up for merging the whole branch, it would still help a lot if you could pull up to before the defconfig change.

popcornmix commented:
@pelwell any objections?

pelwell commented:
I'm skeptical about the need for CONFIG_DEBUG_GPIO, which appears to just enable some dmesg KERN_DEBUG output. AFAICT the /sys/kernel/debug/pinctrl/ entries are enabled by CONFIG_DEBUG_FS, which is already defined. I am, of course, open to correction.

anholt commented:
Yeah, I must have screwed something up to write that commit.  I'll drop it and retest, or you can just drop it.

I'm working on a build against the current rpi-4.4.y right now, and if it works I'll re-push.

pelwell commented on drivers/pinctrl/bcm/pinctrl-bcm2835.c:
```diff
@@ -307,7 +307,7 @@ static inline void bcm2835_pinctrl_fsel_set(
 	enum bcm2835_fsel cur = (val >> FSEL_SHIFT(pin)) & BCM2835_FSEL_MASK;
 
-	dev_dbg(pc->dev, "read %08x (%u => %s)\n", val, pin,
+	dev_info(pc->dev, "read %08x (%u => %s)\n", val, pin,
```
Hmm - I don't think we want these.

pelwell commented on arch/arm/boot/dts/bcm2708_common.dtsi:
```diff
@@ -132,6 +134,11 @@
 			interrupt-controller;
 			#interrupt-cells = <2>;
+
+			i2c0_gpio28: i2c0_gpio28 {
```
It bothers me that this node isn't referenced anywhere. Beyond that, we would have to check that the firmware isn't going to dynamically reconfigure the function mapping on these pins, since i2c0 is a.k.a. i2c_vc.

pelwell commented:
Cool - it's nice to have my sanity confirmed occasionally,

Aside from that and the few inline comments, I have no issues with the patch., but note that I've concentrated on the interactions with other areas rather than reading all of your code.

anholt commented:
Thanks for the cleanups @pelwell.  Pushed with those two fixed and the gpio debug config dropped.

anholt commented:
Also, anyone have any ideas why the panel built as a module wouldn't automatically load?  That's the prereq for dropping CONFIG_DRM_MIPI_DSI to a module, and thus CONFIG_DRM back to module.

pelwell commented:
Off the top of my head (which is going to bed now), is it because the parent of the touchscreen node doesn't cause subnodes to be treated as potential platform devices, perhaps because it isn't a bus?

anholt commented:
Yeah, they're not instantiated as platform devices, they're on their own bus.  I may have a fix.

anholt committed:
of: Export of_device_uevent_modalias for DRM_MIPI_DSI.

In order to make DSI panel modules able to load automatically, we need
to emit the OF modalias uevent so that their MODULE_DEVICE_TABLE(of,
...) can match against it.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
drm: Emit modalias uevents for the DSI devices we create.

This gets the Raspberry Pi panel module to automatically load at boot
time.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt commented:
And fixed a stray debug printf in vc4_dsi.c.

chbae commented:
@anholt
Thanks for this work.
It works fine on GENIVI platform. (https://github.com/genivi/genivi-dev-platform).
I tested on my raspberrypi2 + Official 7'' touch screen. After adding rpi-ft5406-overlay, touch is also working fine.

tom--pollard commented:
Can confirm this is working for me also, thanks for the great work @anholt !
